implemented jsonp support for controllers and added tests

pb82 opened the pull request:
Adds jsonp support to controllers. While CORS is usually preferable for cross domain requests I think there is also some value in having jsonp support: works well with older browsers (IE < 10), good jQuery integration, etc.

The syntax for the jsonp feature is the same as for json:

```elixir
jsonp conn, %{foo: :bar}
```
This will produce the following output on request (GET /?callback=cb):

```javascript
/**/ typeof cb === 'function' && cb({"foo":"bar"});
```

It's possible to override the callback parameter:

```elixir
jsonp conn, %{foo: :bar}, callback: "cb"
```
that allows you to use the overridden callback name in the query string (GET /?cb=cb)

josevalim commented on lib/phoenix/controller.ex:
```diff
@@ -133,6 +133,12 @@ defmodule Phoenix.Controller do
   @spec controller_template(Plug.Conn.t) :: binary
   def controller_template(conn), do: conn.private[:phoenix_template]
 
+  @compile {:inline, get_json_encoder: 0}
```
There is no need to inline it.

pb82 replied:
I thought that since `get_json_encoder` is both small and (potentially) called frequently it would make a good candidate for inlining.

josevalim commented on lib/phoenix/controller.ex:
```diff
@@ -146,14 +152,57 @@ defmodule Phoenix.Controller do
+
+      # Restrict callback name to alphanumeric characters, $, _, . and 
+      # square brackets
+      callback = Regex.replace(~r/[^\[\]\w$.]/, callback, "")
```
I would rather raise if the format is not valid. We should avoid regexes too.

josevalim replied:
We should not even support dots. See here: http://www.benhayak.com/2015/05/stealing-private-photo-albums-from-Google.html

I would keep it to alphanumeric and underscore. No square brackets too as `window[variable]` can be a vector too.

josevalim commented on lib/phoenix/controller.ex:
```diff
@@ -146,14 +152,57 @@ defmodule Phoenix.Controller do
+  
+      # Replace unicode characters that are valid in json but not
+      # in javascript (line separator, paragraph separator)
+      body = encoder.encode!(data)
```
encode! is not part of the format_encoder contract. You should call encoder_to_iodata! and then use `IO.iodata_to_binary`.

josevalim commented on test/phoenix/controller/controller_test.exs:
```diff
@@ -132,6 +132,45 @@ defmodule Phoenix.Controller.ControllerTest do
+  end
+    
+  test "jsonp/3 allows to override the callback name" do
+    conn = conn(:get, "/?cb=cb") |> fetch_query_params()
```
The callback name is the same as the one below. :)

josevalim commented:
Thank you. I have added some comments!

chrismccord commented on lib/phoenix/controller.ex:
```diff
@@ -146,14 +152,57 @@ defmodule Phoenix.Controller do
+    callback_param = Keyword.get(opts, :callback, "callback")
+    callback = Dict.get(conn.query_params, callback_param)
+
+    unless callback do
```
We should avoid unless...else because it's hard to understand. @josevalim since when is unless/else allowed? I didn't think it was valid and was glad when I thought that was the case.

doomspork replied:
Positive case first, :+1: for this suggestion.  Would be awesome to have an Elixir styleguide or best practices document, don't believe there is one currently.

josevalim replied:
It was always allowed. But yes, unless/else is asking for trouble.

pb82 replied:
I will rewrite that using `case` :smirk:

josevalim replied:
`if` would also be fine. The trouble is just the double negation.

pb82 commented:
Thank you for the review @josevalim and @chrismccord. I will update my pull request as soon as possible and take your comments into account.

pb82 commented:
I've updated the pull request: 

* removed the inlining.
* no more double negation.
* only alphanumeric and underscore allowed for callback names.
* no more regex use. I'm not sure if my solution is optimal though.

What do you think?

josevalim commented on lib/phoenix/controller.ex:
```diff
@@ -146,14 +151,66 @@ defmodule Phoenix.Controller do
+    callback_param = Keyword.get(opts, :callback, "callback")
+    callback = Dict.get(conn.query_params, callback_param)
+
+    if callback && String.length(callback) > 0 do
```
`String.length` is going to traverse the whole string. You want to use `byte_size`, which is `O(1)`, or just check `callback != ""`.

pb82 replied:
Ah yes, you're right :) I thought that i can't do this because callback may be `nil`. But of course i can set the default value of `Dict.get` to the empty string. Will do that!

josevalim commented on lib/phoenix/controller.ex:
```diff
@@ -146,14 +151,66 @@ defmodule Phoenix.Controller do
+    if callback && String.length(callback) > 0 do
+      encoder = get_json_encoder()
+
+      allowed_chars = [?0..?9, ?A..?Z, ?a..?z, [?_]]
```
You could have a validate_callback! function:

```elixir
def validate_callback!(<<h, t::binary>>) when h in ?0..?9 or h in ?A..?Z or h in ?a..?z or h == ?_, do: validate_callback!(t)
def validate_callback!(<<>>), do: :ok
def validate_callback!(_), do: raise(ArgumentError, "invalid callback ......")
```

josevalim commented on lib/phoenix/controller.ex:
```diff
@@ -146,14 +151,66 @@ defmodule Phoenix.Controller do
+
+      send_resp(conn, conn.status || 200, "text/javascript", body)
+    else
+      json conn, data
```
Maybe we should raise if there is no callback? What is the purpose of returning json for a jsonp call? maybe it could even be a security issue?

pb82 replied:
The advantage is that it let's you define an API that works both, same domain and cross domain. When it's not cross domain you just leave away the callback and get a normal json response. Otherwise you would have to treat those two cases separately.

josevalim replied:
Ah, I see, perfect. Thank you!

josevalim commented on lib/phoenix/controller.ex:
```diff
@@ -146,14 +151,66 @@ defmodule Phoenix.Controller do
+      |> IO.iodata_to_binary()
+
+      body = body
+      |> String.replace(<< 0x2028 :: utf8 >>, "\\u2028")
```
Nitpick (sorry!): no spaces after `<<` and before `>>`

josevalim commented:
I've added a couple more comments and one small nitpick. :)

jeregrine commented on lib/phoenix/controller.ex:
```diff
@@ -146,14 +151,66 @@ defmodule Phoenix.Controller do
+
+      callback_valid || raise ArgumentError, "the callback name contains invalid characters"
+
+      body = encoder.encode_to_iodata!(data) 
```
could you just call encoder.encode! here and skip the IO.iodata_to_binary?

https://github.com/devinus/poison/blob/master/lib/poison.ex#L39

josevalim replied:
I asked him to not call encode! because it is not part of our  `format_encoder` contract. We just call `encode_to_iodata!`.

jeregrine commented:
Very cool!

pb82 commented:
Thank you, your comments are really useful! I updated the pull request and took the freedom to add your suggested validation function (just made it private) ;)

josevalim commented:
This looks perfect, thank you. @chrismccord any final remarks?

chrismccord commented:
<3 <3 <3
